pex/build_system/pep_518.py

```diff
@@ -0,0 +1,183 @@
+
+if TYPE_CHECKING:
+    from typing import Iterable, Mapping, Optional, Text, Tuple, Union
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+def _execute_with_toml(
+    resolver,  # type: Resolver
+    script_contents,  # type: str
+):
+    # type: (...) -> Union[Text, Error]
+    target_dir = os.path.join(ENV.PEX_ROOT, "tools", "toml.pex", isolated().pex_hash)
+    with TRACER.timed("Building toml parsing PEX"):
+        with atomic_directory(target_dir=target_dir, exclusive=True) as atomic_dir:
+            if not atomic_dir.is_finalized():
+                lock_file_contents = pkgutil.get_data(__name__, "toml-lock.json")
+                assert lock_file_contents is not None, (
+                    "The sibling resource toml-lock.json of {} should always be present in a "
+                    "Pex distribution or source tree.".format(__name__)
+                )
+                contents = lock_file_contents.decode("utf-8")
+                source = os.path.join(os.path.dirname(__file__), "toml-lock.json")
+                lock = json_codec.loads(lockfile_contents=contents, source=source)
+
+                # Our toml lock was created against the default public PyPI index; so we can only
+                # use the lock if the resolve is also setup to use the default repos configuration
+                # as well.
+                if resolver.is_default_repos():
+                    result = resolver.resolve_lock(lock=lock)
+                else:
+                    result = resolver.resolve_requirements(
+                        requirements=[str(req) for req in lock.requirements]
+                    )
+
+                pex_builder = PEXBuilder(path=atomic_dir.work_dir)
+                pex_builder.info.pex_root = ENV.PEX_ROOT
+                pex_builder.info.venv = True
+                for installed_distribution in result.installed_distributions:
+                    pex_builder.add_distribution(installed_distribution.distribution)
+                    for direct_req in installed_distribution.direct_requirements:
+                        pex_builder.add_requirement(direct_req)
+                pex_builder.freeze(bytecode_compile=False)
+
+    try:
+        return subprocess.check_output(
+            args=[ensure_venv(PEX(target_dir)), "-c", script_contents]
+        ).decode("utf-8")
+    except CalledProcessError as e:
+        return Error(str(e))
+
+
+@attr.s(frozen=True)
+class BuildSystemTable(object):
+    requires = attr.ib()  # type: Tuple[str, ...]
+    build_backend = attr.ib()  # type: str
+
+
+def _read_build_system_table(
+    resolver,  # type: Resolver
+    pyproject_toml,  # type: str
+):
+    # type: (...) -> Union[Optional[BuildSystemTable], Error]
+    output_or_error = _execute_with_toml(
+        resolver,
+        dedent(
+            """\
+        import json
+        import sys
+
+        import toml
+
+        with open({pyproject_toml!r}) as fp:
```

Comment:
Presumably there is an obvious reason I'm not seeing, but why not simply parse toml in-process? Is it to avoid introducing a direct dep on the toml library? I ask because a lot of code here is just for running this parsing snippet in a toml pex, and the alternative of vendoring toml would be a lot simpler?

Reply:
I'm on a quest to get rid of all vendoring save for Pip. I've already ~killed setuptools and With the eventual drop of 2.7 & 3.5 support I'll be able to kill attrs. The setuptools and wheel needed to perform legacy builds are better resolved as ~latest. All that said, toml does vendor fine; so I'll do that for now.

Reply:
What's the reason? The vendoring seems pretty solid, and I've started playing around with vendoring portalocker in order to see what it would take to support Windows. Granted, that's not a lot of code and we could copy the salient parts into a single file, but still, what's the objection to vendoring?

Reply:
It barely works (doesn't work for some dists) and Red Baron died in 2019. Pip isn't vendored in the hard way (its imports are not re-written) and so it could just be dropped in w/o Red Baron.

On the portalocker front, we've had two contributors implement that bit without any library. The thing stopping all comers so far is my demand a Windows CI that is green for units & ITs comes along with the production fixes.

Reply:
If you feel that the long-term goal is to get rid of vendoring then I'm fine with this more complex solution, just wanted to verify that it is worth its weight.

Reply:
It's definitely a longer term goal, but I will just vendor for now. A long driving day just finished so I just haven't had time yet.

Reply:
SG just figured since you've already written it and I've already reviewed it, the alternative is also fine

Reply:
Now with toml vendored.